src/nvim/os/stdpaths.c

```diff
@@ -69,13 +69,19 @@ const char *get_appname(void)
   return env_val;
 }
 
-/// Ensure that APPNAME is valid. In particular, it cannot contain directory separators.
+// Ensure that APPNAME is valid. In particular, it cannot contain .. surrounded
+// by path separators, the start of the string, or the end of the string.
 bool appname_is_valid(void)
 {
   const char *appname = get_appname();
   const size_t appname_len = strlen(appname);
+  if (path_is_absolute(appname)) {
+    return false;
+  }
   for (size_t i = 0; i < appname_len; i++) {
-    if (appname[i] == PATHSEP) {
+    if (appname[i] == '.' && appname[i + 1] == '.'
+        && (i == 0 || appname[i - 1] == PATHSEP)
+        && (appname[i + 2] == '\0' || appname[i + 2] == PATHSEP)) {
       return false;
```

Comment:
Can we get rid of the for-loop, and just return false if `..` is found anywhere in the string?

Reply:
This makes sense but would fail for instances of `..dir` or `dir..`. Although these must be very niche cases so I'd be comfortable overlooking them

Reply:
To cover those I could check for 
```
/../
^../
/..$
^..$
```
This is pretty much what the for loop checks for but would be clearer

Reply:
NVIM_APPNAME is a completely controllable value, there's no reason we need to support weird cases. Better for the logic to be simpler where possible.

Reply:
Took me way too long to realise I could use PATHSEPSTR to make this just work.

I also added a check for APPNAME == "." I feel that would break things